### DISCUSSION TITLE
Should we move our custom blocks into separate plugins

### DISCUSSION THREAD

**Discussion by AmnestyAM** [amnestywebsite/humanity-theme#318] [Enhancement Requests]
This would enable users to change theme more easily

**Comment by creativecoder** [comment]
One decision here is whether to use a single plugin that contains all of the Humanity blocks, or distribute each block as an [individual block plugin](https://developer.wordpress.org/plugins/wordpress-org/block-specific-plugin-guidelines/#block-plugins-and-the-block-directory).

### I. One plugin with all Humanity blocks

- This is the easiest to manage and distribute, as it would only be a single plugin to update and maintain.
- The disadvantage is that there are no special integrations in the editor for multi-block plugins like there are for single block plugins (see below).

### II. Each block as a single block plugin

- This will be harder to maintain, as it means updating and deploying each block as its own plugin. This can all be automated with Github actions, but is not trivial to set up.
- Users can install only the blocks they need.
- The editor gives special treatment to single block plugins:

#### One click install for missing blocks

If a site activates the Humanity theme and is using a template with missing blocks, a missing block can be installed and activated in one click from the editor (in the screenshot below, with both blocks unavailable, the Social Sharing single block plugin can be installed and activated in one click, while the Amnesty block cannot).

<img width="659" alt="image" src="https://github.com/user-attachments/assets/6ac0790a-cfa1-459a-92f5-1fd3558f3185">

#### Discovery in the block inserter

Single block plugins appear in the search results of the block inserter, and can be installed, activated, and inserted with one click.

<img width="911" alt="Screenshot 2024-06-05 at 15 15 08" src="https://github.com/user-attachments/assets/8be6f686-a9ab-4f31-b3a5-c77a79a52aad">

---

In both cases, the Humanity theme would need to include information in the readme.txt to indicate installing the plugin is needed to support the templates and patterns that use those blocks. Some day we may be able to declare plugin dependencies for themes (that was originally scoped for, but then cut from the [plugin dependencies work earlier this year](https://make.wordpress.org/core/2024/03/05/introducing-plugin-dependencies-in-wordpress-6-5/))--but that's not possible right now.

**Reply by jaymcp** [reply]
Thanks for the info! 

Maintenance burden is very much a concern, but we could perhaps find a middle ground between the two implementations by using a monorepo, and having the build system produce separate plugins from a single source.

This will probably be a fair amount of work to set up, but it would keep the maintenance burden at a minimum, whilst still providing the advantages of installability via the inserter.

Having said that, I think such an approach would be most fruitful if we are able to provide a mechanism in the theme to suggest/recommend/enable the installation of the Humanity blocks when the theme is installed/activated/updated, so that's something I'll need to think about... assuming that there is no in-built feature for that.

**Reply by creativecoder** [reply]
> Having said that, I think such an approach would be most fruitful if we are able to provide a mechanism in the theme to suggest/recommend/enable the installation of the Humanity blocks when the theme is installed/activated/updated, so that's something I'll need to think about... assuming that there is no in-built feature for that.

Nothing yet that's built in. There's a little bit of discussion about that in this issue: https://github.com/WordPress/gutenberg/issues/62807.
